### PR TITLE
Seek test

### PR DESCRIPTION
**Pull request by daschuer**
This adds some seek tests and make it work with soundsourcemp3 

The changes insoundsourcemp3 are ugly hacks. 
I would prefer to fix this on Uwes new base. 

A backward seek from a late position in the fill still produces a big offset.
IMHO the current state is petty useless for reliable beatmaching :-(

@uklotzde will your new API work with these tests? 
Do you see a chance to fix the seek errors in a way that it will merge well later with our code? 
It is not much fun to to the work twice.

### CODE REVIEW

**Comment by ywwg** [comment]
hey this might actually fix the stuttering problem at the start of vbr mp3s.

**Comment by ywwg** [comment]
(it sounds like it does on a file I tested, but I don't have a good testcase)

**Comment by uklotzde** [comment]
@daschuer I will try integrate and run your new tests on my branch, curiously awaiting the results ;)

**Comment by ywwg** [comment]
> I would prefer to fix this on Uwes new base.

Just as a reminder, we already made it clear that Uwe's change is too large and risky to include in this release.  It's definitely something we will look at post 1.12.

**Comment by uklotzde** [comment]
NewSoundSourceAPI: Failure in seekForward for MP3 when prefetching less than 5 MP3 frames:

pData2[0] = 0.00676334 (kSeekFramePrefetchCount = 0)
pData2[0] = -0.269485 (kSeekFramePrefetchCount = 1)
pData2[0] = -0.264809 (kSeekFramePrefetchCount = 2 / 3 / 4)

Recoverable MP3 decoding error: bad main_data_begin pointer 
Recoverable MP3 decoding error: bad main_data_begin pointer 
"/home/uk/Projects/Mixxx/mixxx/src/test/id3-test-data/cover-test.mp3" Test Sample 0 
src/test/soundproxy_test.cpp:212: Failure
Value of: pData2[i]
  Actual: -0.264809
Expected: pData1[i]
Which is: 0.228699

All tests pass and the "Recoverable MP3 decoding error: bad main_data_begin pointer " warnings disappear when prefetching 5 or more MP3 frames after seeking! Therefore I will increase the prefetch count from 2 to 5 in my branch.

I expect similarities between the code for decoding MP3 and M4A. Both decode the input data in chunks and need some amount of prefetching after seeking. I'm pretty sure that the new M4A code is correct now. Porting the code for seeking to the MP3 decoder would be interesting. I also need to run these tests on the M4A decoder itself to determine the minimum number of prefetch blocks that are required. It's a shame that this is not documented anywhere!

@daschuer: Theses unit tests are very helpful, great idea! Tomorrow I will look at the changes you made in SoundSourceMp3. I think I've already made substantial changes to Audio-/SoundSourceMp3, even if it is not a clean re-write like I did for Audio-/SoundSourceM4A.

File: src/soundsourcemp3.cpp [hunk 1]
```diff
@@ -515,6 +478,10 @@ unsigned SoundSourceMp3::read(unsigned long samples_wanted, const SAMPLE * _dest
                 break;
             }
         }
+        if (frameBefore == Stream->this_frame) {
+            // No seek, try again
+            continue;
```
Comment (daschuer) [inline src/soundsourcemp3.cpp]: @uklotzde: I am currently debugging, why this is required. It happens always after a seek. 
No result until now :-(
This is probably a bug in our init code, or something that triggers a mad bug.

**Comment by daschuer** [comment]
@uklotzde: It could be great if we where able too backport a common solution for libmad to 1.11.

File: src/soundsourcemp3.cpp (hunk 1 above)
Comment (uklotzde) [inline src/soundsourcemp3.cpp]: This seems to happen in the following situations
* mad_frame_decode != 0
* mad_frame_decode == 0: At the beginning of the stream, i.e. for the first frame

**Comment by uklotzde** [comment]
This link might also be helpful:
http://www.mars.org/mailman/public/mad-dev/2002-May/000634.html

The bad news: In the worst case you need to prefetch up to 29 (MP3) frames

**Comment by uklotzde** [comment]
If you increase kSeekSample from 10000 to 20000 the test fails again.

My modified MP3 code also does not  pass the tests, no matter how many frames I'm prefetching.

**Comment by daschuer** [comment]
I assume that inside MAD are Infinity filters, so that we might never reach exact the same samples after seeking. 
Since you see the full resolution of the mad samples we may need a fuzzy range, that passes the test. 
A sample difference from 1 .. 2 16-bit samples should be not hear able as click. 

Do you think we should actually pre-fetch 29 frames? 
I assume, if we loose some bit from the reservoir, we may get some distortions. That would be OK for me if this is in the fuzzy range defined above and beat is still at the same place.

File: src/soundsourcemp3.cpp (hunk 1 above)
Comment (daschuer) [inline src/soundsourcemp3.cpp]: mad_frame_decode returns 0 in this case, but dos not advance to the next frame. 
If mad_frame_decode return != 0 we continue or break the loop before.

**Comment by uklotzde** [comment]
Unfortunately the sample values differ considerably, covering the whole range [-1.0, 1.0]. I have absolutely no no clue why this happens, even when prefetching more than 29 MP3 frames after seeking!? Maybe some kind of jitter?

I've extended your tests by adding an outer loop to cover the whole range of seek offsets starting from 0 to the end of the file with a constant step size (= 10000).

**Comment by daschuer** [comment]
I have made now my test comparable to yours, And it turns out that the seek version has lost one frame when the test breaks. The read version is at frame 43 and the seek version is at 44.
This happens at seekFrameIndex = 50000 / 1152 = 43,4. So the seek version is one ahead as if has skipped one of the prefetch frames.  

The last working is at seekFrameIndex 40000 / 1152 = 34 which works fine

**Comment by daschuer** [comment]
Seek test works now, However, I got some funny debug output.
My test track from the original bug report works like charm now. 
Waveform and Audio is in perfect sync.

Debug output 

seekFrameIndex = 1190000 
MAD: No seek, try again 
MAD: Recoverable frame level ERR ( bad main_data_begin pointer ) 
MAD: No seek, try again 
MAD: No seek, try again 
MAD: Recoverable frame level ERR ( bad main_data_begin pointer ) 
MAD: skip frame 
skip!! 
MAD: Recoverable frame level ERR ( bad main_data_begin pointer ) 
MAD: skip frame 
skip!! 

This explains the issue a bit:
http://permalink.gmane.org/gmane.comp.audio.mad.devel/647

###

I have also noticed that this code always seeks during normal playback. This should be catched,
otherwise we have always the 29 frames = 0,4 s penalty.

**Comment by daschuer** [comment]
@uklotzde: How should we continue? I am not willing to bring this code into a good shape, when I know that the work is already done in a very much better way.

So I think the best way is to backport your changes to master. Once you have integrated the fix in your branch, I can give it a try and do the back port. The biggest question for me is on which level. 
We have the full range from riping of the inner part of your functions over a glue layer up to a fully merge. 
For a fully merge, it looks like we are missing tester on all platforms, with a good variation of all file  types.

**Comment by uklotzde** [comment]
Congratulations, Daniel :) I'm still struggling with the accurate seeking code. Even if it sounds ok for the test file, the test fails miserably. I already fetched your branch and will do another comparison to find the differences.

**Comment by uklotzde** [comment]
Awesome, that code really works! It's all about the correct interpretation of mad_frame_decode()'s result.

I will prepare a backport for SoundSourceMp3 with some of the improvements I already made for the new API.

**Comment by daschuer** [comment]
Cool, Thank you very much! :-)

**Comment by uklotzde** [comment]
The backport will include the following improvements:
* uses fast O(log n) binary searching to find MP3 frames when seeking
* avoids unnecessary seeks: only seeks when moving backward or when jumping more than the number of prefetch frames forward
* ...

**Comment by uklotzde** [comment]
Backport almost done, but I did not have enough time to complete it. First I want to analyze the logs from last weekend, several ours of playing time using the new API without any major issues :) Actually none that are related to audio decoding and the new API.

**Comment by daschuer** [comment]
Thank you, I am patient.  Please do not stress yourself too much ;-)

**Comment by uklotzde** [comment]
I've pushed the branch 'seek_test_backportmp3' to my repository:

https://github.com/uklotzde/mixxx/tree/seek_test_backportmp3

Files changed compared to your 'seek_test' branch: soundsourcemp3.h/.cpp

But be warned, the changes are quite substantial compared to the original version ;) I'm not planning any further work on this, because keeping both versions aligned is time-consuming.

The backport version is even more complicated than the new API version, because it needs to distinguish between the actual number of channels in the file and the virtual number of channels (always 2) that are exposed by getChannels().

**Comment by uklotzde** [comment]
This PR should be obsolete now that https://github.com/mixxxdj/mixxx/pull/411 has been merged to master. I have adopted Daniel's unit tests to verify that seeking works as expected for all supported audio formats.
